engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/BuiltinObject.java

```diff
@@ -0,0 +1,108 @@
+import org.enso.interpreter.runtime.data.EnsoObject;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+
+/**
+ * Base class for every Enso builtin object. Not type. Note that base class for a builtin type is
+ * {@link Builtin}.
+ *
+ * <p>In other words, this class represents an object of builtin type in a similar way that {@link
+ * org.enso.interpreter.runtime.data.atom.Atom} represents an object of a non-builtin type.
+ */
+@ExportLibrary(InteropLibrary.class)
+@ExportLibrary(TypesLibrary.class)
+public abstract class BuiltinObject extends EnsoObject {
+
+  @ExportMessage
+  public final boolean hasType() {
+    return true;
+  }
+
+  /**
+   * Returns the name of the builtin as saved inside {@link Builtins#builtinsByName}. Not fully
+   * qualified.
+   *
+   * @return
+   */
+  protected abstract String builtinName();
+
+  protected final Type getBuiltinType() {
+    return GetType.uncached(this);
+  }
+
+  /**
+   * Must return false, otherwise if a builtin object is passed to a host method that has a single
+   * {@code Object} argument, host interop would convert the builtin object to a {@code Map} with
+   * all its members. Even if the builtin object is, e.g., a number of a date.
+   *
+   * <p>Must return false as long as all our stdlib Java methods accept {@code Object} and not
+   * {@link org.graalvm.polyglot.Value} as arguments comming from Enso.
+   */
+  @ExportMessage
+  public final boolean hasMembers() {
+    return false;
+  }
+
+  @ExportMessage
+  public final Object getMembers(boolean includeInternal) throws UnsupportedMessageException {
+    throw UnsupportedMessageException.create();
+  }
+
+  @ExportMessage
+  public final boolean hasMetaObject() {
+    return true;
+  }
+
+  @ExportMessage(name = "getType", library = TypesLibrary.class)
+  @ExportMessage(name = "getMetaObject", library = InteropLibrary.class)
+  public static class GetType {
+
+    /**
+     * Caching on class of the receiver - as long as there is the same class, its {@link
+     * #builtinName()} method will return the same value. Note that we don't want to cache on the
+     * builtin name, as that would create a separate polymorph cache for every instance of the
+     * receiver.
+     */
+    @Specialization(
+        guards = {"cachedReceiverClass == receiver.getClass()", "getCtx(node) == cachedCtx"},
+        limit = "1")
+    public static Type doItCached(
+        BuiltinObject receiver,
+        @Bind("$node") Node node,
+        @Cached("receiver.getClass()") Class<? extends BuiltinObject> cachedReceiverClass,
+        @Cached(value = "getCtx(node)", allowUncached = true) EnsoContext cachedCtx,
+        @Cached(value = "getBuiltinType(receiver, cachedCtx)", allowUncached = true)
+            Builtin cachedBuiltinType) {
+      return cachedBuiltinType.getType();
+    }
+
+    @Fallback
+    public static Type uncached(BuiltinObject receiver) {
+      var ctx = getCtx(null);
```

Comment:
`getCtx(null)` should rather be avoided. Do we really need `@Fallback` when `doItCached` specialization has `allowUncached = true`? It should work without `@Fallback` somehow in my opinion.

Reply:
> Do we really need @Fallback when doItCached specialization has allowUncached = true? It should work without @Fallback somehow in my opinion.

We probably technically don't need `@Fallback`, but it is useful for https://github.com/enso-org/enso/blob/7a2ea7c0a5ecb45c3b6130ca82b39bd9054f8b97/engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/BuiltinObject.java#L43-L45 which is called from a subclass.

I have at least refactored such that there is not a call `getCtx(null)` but `getCtx(node)` in 32bd04218411feba46055640b2248e0e2c59f9b1